File: backend/app/core/export_safety.py
"""Ochrona eksportów CSV/XLSX przed wstrzyknięciem formuły (CSV injection).

Tekst w eksportach pochodzi od ludzi — także anonimowych: imię, nazwisko
i lokalizacja kandydata przychodzą z publicznego formularza strony kariery,
nazwy klientów i numery zamówień z importów i maili. Komórka zaczynająca się
od ``=`` (albo ``+``, ``-``, ``@``) jest w Excelu formułą: ``=HYPERLINK(…)``
w polu „imię” zamienia plik w link do cudzego serwera, a openpyxl zapisuje
taki napis wprost jako formułę.

Reguła OWASP: tekst zaczynający się od ``= + - @``, tabulatora albo powrotu
karetki dostaje prefiks ``'`` — Excel i LibreOffice traktują wtedy komórkę jako
tekst. Liczby, daty i ``None`` przechodzą bez zmian (zostają liczbami i datami
w arkuszu).

Każdy moduł w ``app/`` budujący CSV/XLSX z danych użytkownika woła
:func:`safe_cell` (pilnuje tego ``tests/test_export_safety_contract.py``).
"""

from __future__ import annotations

from typing import Any, Iterable

# `\t` i `\r` na początku też uruchamiają interpretację w części arkuszy
# (OWASP „CSV Injection”), więc traktujemy je jak znak formuły.
FORMULA_PREFIXES: tuple[str, ...] = ("=", "+", "-", "@", "\t", "\r")


def safe_cell(value: Any) -> Any:
    """Zwróć wartość bezpieczną do komórki arkusza.

    Tylko tekst jest zmieniany — liczby i daty zostają typami, które arkusz
    umie sortować i sumować.
    """

    if isinstance(value, str) and value.startswith(FORMULA_PREFIXES):
        return f"'{value}"
    return value


def safe_row(values: Iterable[Any]) -> list[Any]:
    """:func:`safe_cell` dla każdej komórki wiersza."""

    return [safe_cell(value) for value in values]

File: frontend/src/lib/export-safety.ts
/**
 * Ochrona komórek CSV/XLSX budowanych W PRZEGLĄDARCE przed wstrzyknięciem
 * formuły (runda 6 audytu) — lustro `backend/app/core/export_safety.safe_cell`.
 *
 * Tekst zaczynający się od `= + - @`, tabulatora albo powrotu karetki Excel
 * i LibreOffice czytają jako formułę (`=HYPERLINK(…)` w nazwisku zamienia
 * plik w link do cudzego serwera), więc dostaje prefiks `'`. Wyjątek jak
 * w backendzie: same cyfry i separatory („+48 600 100 200”, „-1 200,50”) nie
 * wywołają funkcji, a apostrof psułby kolumnę telefonu. Liczby i puste
 * wartości przechodzą bez zmian.
 */

const FORMULA_PREFIXES = ["=", "+", "-", "@", "\t", "\r"] as const;
const NUMBER_LIKE = /^[+-][\d\s().,/-]*\d[\d\s().,/-]*$/;

export function safeSpreadsheetCell<T>(value: T): T | string {
  if (typeof value !== "string") return value;
  if (!FORMULA_PREFIXES.some((prefix) => value.startsWith(prefix))) return value;
  if (NUMBER_LIKE.test(value)) return value;
  return `'${value}`;
}
